lib/iris/analysis/__init__.py

```diff
@@ -506,22 +528,270 @@ def aggregate(self, data, axis, **kwargs):
+
+        """
+        if isinstance(data_result, biggus.Array):
+            collapsed_cube.lazy_data(data_result)
+        else:
+            collapsed_cube.data = data_result
+
+        return collapsed_cube
+
+    def aggregate_shape(self, **kwargs):
+        """
+        The shape of the additive dimension created by the aggregator.
+
+        Kwargs:
+
+        * This function is intended to be used in conjunction with aggregate()
+          and should be passed the same keywords.
+
+        Returns:
+            A tuple of the additive dimension shape.
+
+        """
+        return ()
+
+    @abstractmethod
+    def name(self):
+        """
+        Returns the name of the aggregator.
+
+        """
+
+
+class AdditiveAggregator(_Aggregator):
+    """
+    The :class:`AdditiveAggregator` class provides common additive aggregation
+    functionality.
+
+    An additive aggregator *may* introduce a new dimension to the data for the
+    statistic being calculated, but only if more than one point is requested.
+    For example, calculating the 50th and 90th percentile will result in a new
+    data dimension with an extent of 2, for each of the quantiles calculated.
+
+    """
+
+    def __init__(self, name, call_func, required, units_func=None,
+                 lazy_func=None, **kwargs):
+        """
+        Create an additive aggregator for the given :data:`call_func`.
+
+        Args:
+
+        * name (string):
+            The name of the aggregator.
+
+        * call_func (callable):
+            | *Call signature*: (data, axis=None, \**kwargs)
+
+            Data aggregation function.
+            Returns an aggregation result, collapsing the 'axis' dimension of
+            the 'data' argument.
+
+        * required (string):
+            The name of the mandatory keyword argument associated with
+            calling this aggregator.
+
+        Kwargs:
+
+        * units_func (callable):
+            | *Call signature*: (units)
+
+            If provided, called to convert a cube's units.
+            Returns an :class:`iris.units.Unit`, or a
+            value that can be made into one.
+
+        * lazy_func (callable or None):
+            An alternative to :data:`call_func` implementing a lazy
+            aggregation. Note that, it need not support all features of the
+            main operation, but should raise an error in unhandled cases.
+
+        Additional kwargs::
+            Passed through to :data:`call_func` and :data:`lazy_func`.
+
+        Aggregators are used by cube aggregation methods such as
+        :meth:`~iris.cube.Cube.collapsed` and
+        :meth:`~iris.cube.Cube.aggregated_by`.  For example::
+
+            cube.collapsed('longitude', iris.analysis.PERCENTILE, percent=50)
+
+        """
+        self._name = name
+        if not isinstance(required, basestring) or not required:
+            msg = '{} aggregator requires a mandatory keyword argument.'
```

Comment:
We dug that hole a long, long time ago ... optional mandatory kwargs, *sigh* which were only *mentioned* in documentation

Really, the users shouldn't see this, it's only at aggregator creation time i.e. when the aggregators are created in `iris/analysis/__init__.py`

The use case for the `AdditiveAggregator` is `PERCENTILE` which must be passed the `percent` kwarg when called by the user. So here I'm being strict and explicit, saying that an `AdditiveAggregator` must have a mandatory kwarg.

So to answer your question, no the user should not see this, most likely it would be a developer or someone who knows what they're doing by creating a new aggregator.

Such a mandatory keyword argument must be passed as a keyword argument when using the aggregator via the usual cube methods `collapsed`, `aggregated_by` and `rolling_window`.